src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -28,56 +28,68 @@ seastar::future<> ReplicatedRecoveryBackend::recover_object(
   // start tracking the recovery of soid
   return seastar::do_with(std::map<pg_shard_t, PushOp>(), get_shards_to_push(soid),
     [this, soid, need](auto& pops, auto& shards) {
-    return maybe_pull_missing_obj(soid, need).then([this, soid](bool pulled) {
-      return load_obc_for_recovery(soid, pulled);
-    }).safe_then([this, soid, need, &pops, &shards] {
-      if (!shards.empty()) {
-	return prep_push(soid, need, &pops, shards);
-      } else {
-	return seastar::now();
-      }
-    }, crimson::ct_error::all_same_way([this, soid](const std::error_code& e) {
-      auto recovery_waiter = recovering.find(soid);
-      if (auto obc = recovery_waiter->second.obc; obc) {
-        obc->drop_recovery_read();
-      }
-      recovering.erase(recovery_waiter);
-      return seastar::make_exception_future<>(e);
-    })).then([this, &pops, &shards, soid] {
-      return seastar::parallel_for_each(shards,
-	[this, &pops, soid](auto shard) {
-	auto msg = make_message<MOSDPGPush>();
-	msg->from = pg.get_pg_whoami();
-	msg->pgid = pg.get_pgid();
-	msg->map_epoch = pg.get_osdmap_epoch();
-	msg->min_epoch = pg.get_last_peering_reset();
-	msg->set_priority(pg.get_recovery_op_priority());
-	msg->pushes.push_back(pops[shard->first]);
-	return shard_services.send_to_osd(shard->first.osd, std::move(msg),
-					  pg.get_osdmap_epoch()).then(
-	  [this, soid, shard] {
-	  return recovering.at(soid).wait_for_pushes(shard->first);
+    return maybe_pull_missing_obj(soid, need).then(
+      [this, soid, need, &pops, &shards](bool pulled) {
+      auto func = [this, soid, need, &pops, &shards] {
```

Comment:
instead of having a large function, probably it'd be easier to future readers if we could extract a separated function which `maybe_push_shards()`? this function just

1. prepare for push op
2. push shard for all pg shard which need to be recovered
3. notify the potential waiters waiting for this previously unavailable object